### PR TITLE
Add the `countdown` argument to minetest.request_shutdown

### PR DESCRIPTION
**Pull request by zaoqi**
Add compact, short information about your PR for easier understanding:

- Goal of the PR : Add the `countdown_msg` argument to minetest.request_shutdown
- How does the PR work?
- Does it resolve any reported issue? no
- If not a bug fix, why is this PR needed? What usecases does it solve? custom countdown message

## To do

This PR is a Ready for Review.

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -4912,12 +4912,13 @@ Timing
...
 ------
 
-* `minetest.request_shutdown([message],[reconnect],[delay])`: request for
+* `minetest.request_shutdown([message],[reconnect],[delay],[countdown])`: request for
   server shutdown. Will display `message` to clients.
     * `reconnect` == true displays a reconnect button
     * `delay` adds an optional delay (in seconds) before shutdown.
       Negative delay cancels the current active shutdown.
       Zero delay triggers an immediate shutdown.
+    * `countdown` message displayed during countdown
```
Comment (sfan5) [inline doc/lua_api.txt]: the name is misleading like this, should be renamed to e.g. `countdown_msg`

File: src/server.cpp [hunk 2]
```diff
@@ -3669,7 +3669,7 @@ void Server::requestShutdown(const std::string &msg, bool reconnect, float delay
 	// Positive delay, tell the clients when the server will shut down
 		std::wstringstream ws;
 
-		ws << L"*** Server shutting down in "
+		ws << countdown
 				<< duration_to_string(myround(delay)).c_str()
 				<< ".";
```
Comment (sfan5) [inline src/server.cpp]: It would be useful if this could be formatted instead of being hardcoded as "Text <seconds>."

File: doc/lua_api.txt (hunk 1 above)
Comment (zaoqi) [reply doc/lua_api.txt]: renamed

File: src/server.cpp (hunk 2 above)
Comment (zaoqi) [reply src/server.cpp]: done.

**Comment by SmallJoker** [comment]
I think it is time for a structural change in the code. Currently there are four arguments for each shutdown-related function.
```C++
			bool is_requested = false;
			bool should_reconnect = false;
			std::string countdown_message;
			std::string message;
```
To improve readability, combine those to a `struct` like this:
```C++
			struct ShutdownInformation {
				bool is_requested = false;
				bool should_reconnect = false;
				std::string countdown_message;
				std::string message;
			};
```
Afterwards you can initialize it once, and re-use in the relevant functions like so:
```C++
void Server::requestShutdown(const Server::ShutdownInformation &info)
{
	.....
	m_shutdown_state.trigger(info);
```

**Comment by zaoqi** [comment]
@SmallJoker changed

File: src/server.cpp [hunk 3]
```diff
@@ -3665,19 +3670,16 @@ void Server::requestShutdown(const std::string &msg, bool reconnect, float delay
 		SendChatMessage(PEER_ID_INEXISTENT, ws.str());
 		// m_shutdown_* are already handled, skip.
 		return;
-	} else if (delay > 0.0f) {
+	} else if (info.delay > 0.0f) {
 	// Positive delay, tell the clients when the server will shut down
+		m_shutdown_state.trigger(info);
```
Comment (sfan5) [inline src/server.cpp]: why was this moved?

File: src/server.h [hunk 4]
```diff
@@ -366,16 +372,13 @@ class Server : public con::PeerHandler, public MapEventReceiver,
 		friend class TestServerShutdownState;
 		public:
 			bool is_requested = false;
-			bool should_reconnect = false;
-			std::string message;
+			ShutdownInformation info{.should_reconnect = false};
```
Comment (sfan5) [inline src/server.h]: spacing

File: src/unittest/test_server_shutdown_state.cpp [hunk 5]
```diff
@@ -68,55 +68,65 @@ void TestServerShutdownState::testInit()
...
+	UASSERT(ss.info.message.empty());
+	UASSERT(ss.info.countdown_message.empty());
+	UASSERT(ss.info.delay == 0.0f);
 }
 
 void TestServerShutdownState::testTrigger()
 {
 	Server::ShutdownState ss;
-	ss.trigger(3.0f, "testtrigger", true);
+	Server::ShutdownInformation info{.should_reconnect = true,
```
Comment (sfan5) [inline src/unittest/test_server_shutdown_state.cpp]: spacing / code style

File: src/unittest/test_server_shutdown_state.cpp [hunk 6]
```diff
@@ -68,55 +68,65 @@ void TestServerShutdownState::testInit()
...
+	UASSERT(ss.info.message == "testtrigger");
+	UASSERT(ss.info.delay == 3.0f);
 }
 
 void TestServerShutdownState::testTick()
 {
 	std::unique_ptr<FakeServer> fakeServer(new FakeServer());
 	Server::ShutdownState ss;
-	ss.trigger(28.0f, "testtrigger", true);
+	Server::ShutdownInformation info{.should_reconnect = true,
```
Comment (sfan5) [inline src/unittest/test_server_shutdown_state.cpp]: same

File: src/server.cpp (hunk 3 above)
Comment (zaoqi) [reply src/server.cpp]: ```m_shutdown_state.getShutdownTimerMessage();``` depends on this line.I moved this line.

File: src/server.h (hunk 4 above)
Comment (zaoqi) [reply src/server.h]: What's correct spacing?

**Comment by zaoqi** [comment]
What's correct spacing?

File: src/server.h (hunk 4 above)
Comment (SmallJoker) [reply src/server.h]: Good question. The spacing rules within brackets are only mentioned in the Lua code style. I'd imagine either
`ShutdownInformation info {.should_reconnect = false};`
or
```C++
ShutdownInformation info {
	.should_reconnect = false
};
```

File: src/unittest/test_server_shutdown_state.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/unittest/test_server_shutdown_state.cpp]: @zaoqi 
```C++
Server::ShutdownInformation info {
	.should_reconnect = true,
	.countdown_message = "testcountdown @1.",
...
};
```

File: src/unittest/test_server_shutdown_state.cpp (hunk 6 above)
Comment (zaoqi) [reply src/unittest/test_server_shutdown_state.cpp]: fixed

File: src/unittest/test_server_shutdown_state.cpp (hunk 5 above)
Comment (zaoqi) [reply src/unittest/test_server_shutdown_state.cpp]: fixed

File: src/server.h (hunk 4 above)
Comment (zaoqi) [reply src/server.h]: fixed

File: doc/lua_api.txt [hunk 7]
```diff
@@ -4915,12 +4915,13 @@ Timing
...
 ------
 
-* `minetest.request_shutdown([message],[reconnect],[delay])`: request for
+* `minetest.request_shutdown([message],[reconnect],[delay],[countdown_msg])`: request for
   server shutdown. Will display `message` to clients.
     * `reconnect` == true displays a reconnect button
     * `delay` adds an optional delay (in seconds) before shutdown.
       Negative delay cancels the current active shutdown.
       Zero delay triggers an immediate shutdown.
+    * `countdown_msg` message displayed during countdown
```
Comment (SmallJoker) [inline doc/lua_api.txt]:
```suggestion
    * `countdown_msg` message displayed during countdown
      The placeholder `@1` will be replaced by the countdown time.
```

Comment (zaoqi) [reply doc/lua_api.txt]: handled.

**Comment by zaoqi** [comment]
ping @sfan5

**Comment by pyrollo** [comment]
Gave Clang satisfaction about `test_server_shutdown_state.cpp` but this goes against @sfan5 recommendations. What to do so ?

**Comment by sfan5** [comment]
Github does not permit me to push changes (to add `// clang-format off` btw), so we can't proceed with this PR right now.
<pre>
ERROR: Permission to zaoqi-clone/minetest.git denied to sfan5.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
</pre>

**Comment by rubenwardy** [comment]
Closed due to inactivity
